### ISSUE TITLE
How to log other losses than loss and val_loss?

### ISSUE THREAD

**Issue by mummpl** [matterport/Mask_RCNN#2325]
Hello everyone,

In Tensorboard as well as in a log csv file I am saving via 
`keras.callbacks.CSVLogger('log.csv', append=True, separator=';')`
I only see loss and val_loss. How do I log the other calculated losses like
`loss_names = ["rpn_class_loss", "rpn_bbox_loss",
                      "mrcnn_class_loss", "mrcnn_bbox_loss", "mrcnn_mask_loss"]`?

**Comment by jtk1919** [comment]
The Internal tensorboard gives probelms.  Turn it off in model.py::train()

`callbacks = [
            keras.callbacks.TensorBoard(log_dir=self.log_dir,
                                        histogram_freq=0, write_graph=False, write_images=False),
]`

Then create a custom tensorboard:

`
class ValidationEvaluator(Callback):  
    def __init__(self, val_dataset, validation_log_dir, config, period=5):  
        super(Callback, self).__init__()  
        self.period = period  
        self.batch_size = config.BATCH_SIZE  
        self.validation_data = []  
        self.validation_log_dir = validation_log_dir  
        self.val_writer = tf.summary.FileWriter(self.validation_log_dir)  
        val_gen = data_generator(val_dataset, config, shuffle=False, batch_size=config.BATCH_SIZE)  
        self.validation_data, _ = next(val_gen)  
        print( "Validation data length: ", len(self.validation_data ))  
    #  
    def on_epoch_end(self, epoch, logs={}):  
        print( "\nValidationEvaluator for epoch %d, batch size %d" % (epoch + 1, self.batch_size ) )  
        if (epoch + 1) % self.period == 0:  
            loss, rpn_class_loss, rpn_bbox_loss, class_loss, bbox_loss,  mask_loss = \  
                self.model.evaluate([self.validation_data[0], self.validation_data[1],  
                                     self.validation_data[2], self.validation_data[3],  
                                     self.validation_data[4], self.validation_data[5], self.validation_data[6] ],  
                                    batch_size= self.batch_size)  
            print('\nEpoch %d: loss: %.4f, class_loss: %.4f, mask_loss: %.4f, bbox_loss: %.4f, rpn_class_loss: %.4f, rpn_bbox_loss: %.4f\n' %  
                   (epoch + 1, loss, class_loss, mask_loss, bbox_loss, rpn_class_loss, rpn_bbox_loss))  
            loss_summary = tf.compat.v1.Summary()  
            loss_summary_value = loss_summary.value.add()  
            loss_summary_value.simple_value = loss  
            loss_summary_value.tag = 'loss'  
            self.val_writer.add_summary(loss_summary, epoch + 1)  
            class_loss_summary = tf.compat.v1.Summary()  
            class_loss_summary_value = class_loss_summary.value.add()  
            class_loss_summary_value.simple_value = class_loss  
            class_loss_summary_value.tag = 'class_loss'  
            self.val_writer.add_summary(class_loss_summary, epoch + 1)  
            mask_loss_summary = tf.compat.v1.Summary()  
            mask_loss_summary_value = mask_loss_summary.value.add()  
            mask_loss_summary_value.simple_value = mask_loss  
            mask_loss_summary_value.tag = 'mask_loss'  
            self.val_writer.add_summary(mask_loss_summary, epoch + 1)  
            bbox_loss_summary = tf.compat.v1.Summary()  
            bbox_loss_summary_value = bbox_loss_summary.value.add()  
            bbox_loss_summary_value.simple_value = bbox_loss  
            bbox_loss_summary_value.tag = 'bbox_loss'  
            self.val_writer.add_summary(bbox_loss_summary, epoch + 1)  
            rpn_class_loss_summary = tf.compat.v1.Summary()  
            rpn_class_loss_summary_value = rpn_class_loss_summary.value.add()  
            rpn_class_loss_summary_value.simple_value = rpn_class_loss  
            rpn_class_loss_summary_value.tag = 'rpn_class_loss'  
            self.val_writer.add_summary( rpn_class_loss_summary, epoch + 1)  
            rpn_bbox_loss_summary = tf.compat.v1.Summary()  
            rpn_bbox_loss_summary_value = rpn_bbox_loss_summary.value.add()  
            rpn_bbox_loss_summary_value.simple_value = rpn_bbox_loss  
            rpn_bbox_loss_summary_value.tag = 'rpn_bbox_loss'  
            self.val_writer.add_summary(rpn_bbox_loss_summary, epoch + 1)  
            self.val_writer.flush()  
`

Now add this to the callbacks to run at the end of each epoch in model.py::train():

`# Callbacks  

        validation_evaluator = ValidationEvaluator(val_dataset, self.log_dir, self.config, period=1)  
        callbacks = [  
            keras.callbacks.TensorBoard(log_dir=self.log_dir,  
                                        histogram_freq=0, write_graph=False, write_images=False),  
            keras.callbacks.ModelCheckpoint(self.checkpoint_path,  
                                            verbose=0, save_weights_only=True),  
            validation_evaluator  
        ]`

**Comment by jtk1919** [comment]
############################################################
#  Custom classes and functions
############################################################

class ValidationEvaluator(Callback):
    def __init__(self, val_dataset, validation_log_dir, config, period=5):
        super(Callback, self).__init__()
        self.period = period
        self.batch_size = config.BATCH_SIZE
        self.validation_data = []
        self.validation_log_dir = validation_log_dir
        self.val_writer = tf.summary.FileWriter(self.validation_log_dir)
        val_gen = data_generator(val_dataset, config, shuffle=False, batch_size=config.BATCH_SIZE)
        self.validation_data, _ = next(val_gen)
        print( "Validation data length: ", len(self.validation_data ))
    #
    def on_epoch_end(self, epoch, logs={}):
        print( "\nValidationEvaluator for epoch %d, batch size %d" % (epoch + 1, self.batch_size ) )
        if (epoch + 1) % self.period == 0:
            loss, rpn_class_loss, rpn_bbox_loss, class_loss, bbox_loss,  mask_loss = \
                self.model.evaluate([self.validation_data[0], self.validation_data[1],
                                     self.validation_data[2], self.validation_data[3],
                                     self.validation_data[4], self.validation_data[5], self.validation_data[6] ],
                                    batch_size= self.batch_size)
            print('\nEpoch %d: loss: %.4f, class_loss: %.4f, mask_loss: %.4f, bbox_loss: %.4f, rpn_class_loss: %.4f, rpn_bbox_loss: %.4f\n' %
                   (epoch + 1, loss, class_loss, mask_loss, bbox_loss, rpn_class_loss, rpn_bbox_loss))
            loss_summary = tf.compat.v1.Summary()
            loss_summary_value = loss_summary.value.add()
            loss_summary_value.simple_value = loss
            loss_summary_value.tag = 'loss'
            self.val_writer.add_summary(loss_summary, epoch + 1)
            class_loss_summary = tf.compat.v1.Summary()
            class_loss_summary_value = class_loss_summary.value.add()
            class_loss_summary_value.simple_value = class_loss
            class_loss_summary_value.tag = 'class_loss'
            self.val_writer.add_summary(class_loss_summary, epoch + 1)
            mask_loss_summary = tf.compat.v1.Summary()
            mask_loss_summary_value = mask_loss_summary.value.add()
            mask_loss_summary_value.simple_value = mask_loss
            mask_loss_summary_value.tag = 'mask_loss'
            self.val_writer.add_summary(mask_loss_summary, epoch + 1)
            bbox_loss_summary = tf.compat.v1.Summary()
            bbox_loss_summary_value = bbox_loss_summary.value.add()
            bbox_loss_summary_value.simple_value = bbox_loss
            bbox_loss_summary_value.tag = 'bbox_loss'
            self.val_writer.add_summary(bbox_loss_summary, epoch + 1)
            rpn_class_loss_summary = tf.compat.v1.Summary()
            rpn_class_loss_summary_value = rpn_class_loss_summary.value.add()
            rpn_class_loss_summary_value.simple_value = rpn_class_loss
            rpn_class_loss_summary_value.tag = 'rpn_class_loss'
            self.val_writer.add_summary( rpn_class_loss_summary, epoch + 1)
            rpn_bbox_loss_summary = tf.compat.v1.Summary()
            rpn_bbox_loss_summary_value = rpn_bbox_loss_summary.value.add()
            rpn_bbox_loss_summary_value.simple_value = rpn_bbox_loss
            rpn_bbox_loss_summary_value.tag = 'rpn_bbox_loss'
            self.val_writer.add_summary(rpn_bbox_loss_summary, epoch + 1)
            self.val_writer.flush()

**Comment by jtk1919** [comment]
Now you can view the loss graphs on Tensorboard:
![Screenshot from 2020-10-18 13-39-22](https://user-images.githubusercontent.com/8102382/96367772-c5929400-1147-11eb-8eb0-3c7ccb9d3bcd.png)

tensorboard --logdir ............/logs/ --host localhost

Open in browser: http://localhost:6006/
